### PR TITLE
refactor: move web3 hooks into it's own package

### PR DESCRIPTION
**Pull request by Jack-Works**

### CODE REVIEW

**Comment by Jack-Works** [comment]
Following function need test cc @AinuX thanks:

- Dashboard: Backup wallet with the mnemonic and private key in hex
- Dashboard: Import wallet
- Change selected wallet
- Change selected ETH chain
- An arbitrary plugin in any chain

**Comment by Jack-Works** [comment]
No need to test for now. Still having some problems

File: packages/maskbook/src/web3/hooks/useChainId.ts [hunk 1]
```diff
@@ -1,47 +1 @@
...
-
-/**
- * Retruns true if chain id is available
- */
-export function useChainIdValid() {
-    const unsafeChainId = useUnsafeChainId()
-    const selectedWallet = useWallet()
-    return !Flags.wallet_network_strict_mode_enabled || unsafeChainId === ChainId.Mainnet || !selectedWallet
-}
+export { useChainId, useChainIdValid } from '@dimensiondev/web3-shared'
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/hooks/useChainId.ts:1]: Same as `useAccount`.

File: packages/maskbook/src/web3/hooks/useConstant.ts [hunk 2]
```diff
@@ -1,13 +1 @@
...
-export function useConstant<T extends Web3Constants, K extends keyof T>(constant: T, key: K, chainId?: ChainId) {
-    const chainId_ = useChainId()
-    return getConstant(constant, key, chainId ?? chainId_)
-}
-
-export function useAllConstants<T extends Web3Constants>(constant: T, chainId?: ChainId) {
-    const chainId_ = useChainId()
-    return getAllConstants(constant, chainId ?? chainId_)
-}
+export { useConstant } from '@dimensiondev/web3-shared'
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/hooks/useConstant.ts:1]: Same as `useAccount`.

File: packages/maskbook/src/web3/hooks/useAccount.ts [hunk 3]
```diff
@@ -1,9 +1 @@
-import { useWallet } from '../../plugins/Wallet/hooks/useWallet'
-
-/**
- * Get the address of the default wallet
- */
-export function useAccount() {
-    const wallet = useWallet()
-    return wallet?.address ?? ''
-}
+export { useAccount } from '@dimensiondev/web3-shared'
```
[the remark this answers is not in the document]
Comment (Jack-Works) [reply packages/maskbook/src/web3/hooks/useAccount.ts:1]: I'm aimed to make the change set as small as possible

**Review by guanbinrui — APPROVED** [review]

File: packages/web3-shared/src/hooks/context/shared-states.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    const chain = useSubscription(_.currentChain)
+    if (!allowTestChain) return ChainId.Mainnet
+    return chain
+}
+/** @internal */
+function useChainIDAvailable(wallets: Wallet[]): boolean {
+    const _ = useWeb3Provider()
+    const allowTestChain = useSubscription(_.allowTestChain)
+    const unsafeChainId = useSubscription(_.currentChain)
+    const selectedWallet = wallets
```
Comment (nuanyang233) [inline packages/web3-shared/src/hooks/context/shared-states.ts:35]: It looks like a mistake

Comment (nuanyang233) [reply packages/web3-shared/src/hooks/context/shared-states.ts:35]: The `wallets` look not be **selected** wallet

Comment (Jack-Works) [reply packages/web3-shared/src/hooks/context/shared-states.ts]: fixed

**Review by nuanyang233 — APPROVED** [review]

**Review by guanbinrui — APPROVED** [review]

File: packages/web3-shared/src/hooks/useConstants.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { useMemo } from 'react'
+import { ChainId, useChainId } from './useChainId'
+
+type Primitive = string | number | boolean
```
Comment (UncleBill) [inline packages/web3-shared/src/hooks/useConstants.ts:4]: Should `symbol | bigint | undefined | null`?

```suggestion
type Primitive = string | number | boolean | symbol | bigint | undefined | null 
```

**Review by UncleBill — APPROVED** [review]

**Pull request merged**